### ISSUE TITLE
-Wreturn-type has many false positives

### ISSUE THREAD

**Issue by 54aefcd4-c07d-4252-8441-723563c8826f** [llvm/llvm-project#49759]
|  |  |
| --- | --- |
| Bugzilla Link | [50415](https://llvm.org/bz50415) |
| Version | trunk |
| OS | Linux |
| CC | @DougGregor,@zygoloid |

## Extended Description 
Example  (https://clang.godbolt.org/z/66YqsGEha):

    int f(unsigned i) {
      if (i < unsigned(-1)) return i;
    }
    int g(unsigned i) {
      if (i >= 0) return i;
    }

compiled with clang++ -std=c++17 (no -Wall, no -Wextra) warns with

    warning: non-void function does not return a value in all control paths [-Wreturn-type]

for both functions. 

AFAICT these warnings are false positives.

**Comment by cor3ntin** [comment]
@Sirraide

**Comment by Sirraide** [comment]
Yeah, I think this runs into the same problem of ‘how much constant folding are we willing to implement in the frontend in order to catch things that *technically* always return’; my answer would be ‘none’, because it’s never going to be perfect anyway because that just runs into the halting problem...

**Comment by Sirraide** [comment]
Also, for `unsigned i;` `i >= 0` I’d expect a warning along the lines of ‘comparison always evaluates to `true`’, because I candidly struggle to see what the point of that code is in the first place (even if it’s templated or whatever, if you end up w/ that after instantiation it’s probably a bug honestly...)

**Comment by Sirraide** [comment]
> Is this a realistic test case?
> 
> It seems to me that the original code is suspicious and should be amended: either the author of the code statically knows that the condition is always true, in which case they wrote a dead branch, or they don't know that, in which case they missed a return statement. So producing _some_ warning on this by default doesn't seem obviously unreasonable.

Yeah, this is my opinion basically: I don’t think we should do anything about this particular case because it’s pretty likely to be a bug one way or the other.

**Comment by AaronBallman** [comment]
Resolving as a duplicate of #111509
